File: android/app/src/main/kotlin/ts/sun/tomato_deeplio/TomatoDeeplioPlugin.kt
package ts.sun.tomato_deeplio

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * TomatoDeeplioPlugin
 *
 * Handles:
 * - App Links (Android 12+ verified links)
 * - Custom URI scheme deep links
 * - Initial link extraction (cold start)
 * - Foreground link stream (EventChannel)
 * - Deferred link token persistence (SharedPreferences)
 */
class TomatoDeeplioPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware,
    EventChannel.StreamHandler {

    companion object {
        private const val METHOD_CHANNEL = "tomato_deeplio"
        private const val EVENT_CHANNEL = "tomato_deeplio/events"
        private const val PREFS_NAME = "tomato_deeplio_prefs"
        private const val KEY_DEFERRED_TOKEN = "deferred_token"
    }

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context
    private var activity: Activity? = null
    private var eventSink: EventChannel.EventSink? = null

    // Holds the initial link URL until Dart side requests it
    private var initialLink: String? = null
    // Holds a pending link received while the app is running
    private var pendingLink: String? = null

    // ── FlutterPlugin ──────────────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext

        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    // ── ActivityAware ──────────────────────────────────────────────────────

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        // Extract the initial link from the launch intent
        initialLink = extractLink(binding.activity.intent)
        binding.addOnNewIntentListener { intent ->
            handleNewIntent(intent)
            true
        }
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // ── MethodCallHandler ──────────────────────────────────────────────────

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getInitialLink" -> {
                result.success(initialLink)
                // Clear after first read so it is not replayed
                initialLink = null
            }
            "storeDeferredToken" -> {
                val token = call.argument<String>("token")
                if (token != null) {
                    prefs().edit().putString(KEY_DEFERRED_TOKEN, token).apply()
                }
                result.success(null)
            }
            "getDeferredToken" -> {
                result.success(prefs().getString(KEY_DEFERRED_TOKEN, null))
            }
            "clearDeferredToken" -> {
                prefs().edit().remove(KEY_DEFERRED_TOKEN).apply()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    // ── EventChannel.StreamHandler ─────────────────────────────────────────

    override fun onListen(arguments: Any?, sink: EventChannel.EventSink?) {
        eventSink = sink
        // Flush any link that arrived before the stream was ready
        pendingLink?.let { link ->
            sink?.success(link)
            pendingLink = null
        }
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // ── Private helpers ────────────────────────────────────────────────────

    private fun handleNewIntent(intent: Intent) {
        val link = extractLink(intent) ?: return
        if (eventSink != null) {
            eventSink?.success(link)
        } else {
            pendingLink = link
        }
    }

    private fun extractLink(intent: Intent?): String? {
        if (intent == null) return null
        val action = intent.action
        val data: Uri? = intent.data

        return when {
            // App Link / Universal Link (HTTPS)
            action == Intent.ACTION_VIEW && data != null -> data.toString()
            // Custom URI scheme
            action == Intent.ACTION_VIEW && data != null -> data.toString()
            else -> null
        }
    }

    private fun prefs(): SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
}

File: ios/Classes/TomatoDeeplioPlugin.swift
import Flutter
import UIKit

/// TomatoDeeplioPlugin (iOS / Swift)
///
/// Handles:
/// - Universal Links (HTTPS — requires apple-app-site-association)
/// - Custom URI scheme deep links
/// - Initial link extraction (cold start)
/// - Foreground link stream (EventChannel)
/// - Deferred link token persistence (UserDefaults)
public class TomatoDeeplioPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    // MARK: - Constants

    private static let methodChannelName = "tomato_deeplio"
    private static let eventChannelName  = "tomato_deeplio/events"
    private static let deferredTokenKey  = "tomato_deeplio_deferred_token"

    // MARK: - State

    private var eventSink: FlutterEventSink?
    private var initialLink: String?
    private var pendingLink: String?

    // MARK: - Registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = TomatoDeeplioPlugin()

        let methodChannel = FlutterMethodChannel(
            name: methodChannelName,
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: eventChannelName,
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)

        registrar.addApplicationDelegate(instance)
    }

    // MARK: - FlutterMethodCallDelegate

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getInitialLink":
            result(initialLink)
            initialLink = nil

        case "storeDeferredToken":
            if let args = call.arguments as? [String: Any],
               let token = args["token"] as? String {
                UserDefaults.standard.set(token, forKey: Self.deferredTokenKey)
            }
            result(nil)

        case "getDeferredToken":
            result(UserDefaults.standard.string(forKey: Self.deferredTokenKey))

        case "clearDeferredToken":
            UserDefaults.standard.removeObject(forKey: Self.deferredTokenKey)
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - FlutterStreamHandler

    public func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        self.eventSink = events
        if let pending = pendingLink {
            events(pending)
            pendingLink = nil
        }
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    // MARK: - UIApplicationDelegate hooks

    /// Cold-start via custom URI scheme (e.g. myapp://product/123)
    public func application(
        _ app: UIApplication,
        open url: URL,
        options: [UIApplication.OpenURLOptionsKey: Any] = [:]
    ) -> Bool {
        handleIncomingUrl(url.absoluteString)
        return true
    }

    /// Cold-start via Universal Link (HTTPS)
    public func application(
        _ application: UIApplication,
        continue userActivity: NSUserActivity,
        restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void
    ) -> Bool {
        guard userActivity.activityType == NSUserActivityTypeBrowsingWeb,
              let url = userActivity.webpageURL else { return false }
        handleIncomingUrl(url.absoluteString)
        return true
    }

    // MARK: - Private helpers

    private func handleIncomingUrl(_ urlString: String) {
        if eventSink != nil {
            eventSink?(urlString)
        } else if initialLink == nil {
            initialLink = urlString
        } else {
            pendingLink = urlString
        }
    }
}
